sql/metadata.go

```diff
@@ -0,0 +1,146 @@
+package sql
+
+import (
+	"database/sql"
+	"fmt"
+	"sqlflow.org/gomaxcompute"
+	"strconv"
+	"strings"
+)
+
+type Metadata struct {
+	odpsConfig *gomaxcompute.Config
+	table      string
+	featureMap *featureMap
+	columnInfo *map[string]*columnSpec
+}
+
+func (meta *Metadata) GetColumnInfo(featureColumns map[string][]featureColumn) (map[string]*columnSpec, error) {
+	allKeys := make([]string, 0)
+	for _, fcs := range featureColumns {
+		fields := GetAllKeys(fcs)
+		allKeys = append(allKeys, fields...)
+	}
+	columns, err := meta.GetSparseColumnInfo()
+	if err != nil {
+		fmt.Println(err)
+		return columns, err
+	}
+	denseKeys := make([]string, 0)
+	for _, key := range (allKeys) {
+		_, present := columns[key]
+		if !present {
+			denseKeys = append(denseKeys, key)
+		}
+	}
+	denseColumns, err := meta.GetDenseColumnInfo(denseKeys)
+	if err != nil {
+		fmt.Println(err)
+		return columns, err
+	}
+	for k, v := range denseColumns {
+		columns[k] = v
+	}
+	return columns, nil
+}
+
+func GetAllKeys(fcs [] featureColumn) ([]string) {
+	output := make([]string, 0)
+	for _, fc := range fcs {
+		key := fc.(alpsFeatureColumn).GetKey()
+		output = append(output, key)
+	}
+	return output
+}
+
+func (meta *Metadata) GetDenseColumnInfo(keys []string) (map[string]*columnSpec, error) {
+	output := map[string]*columnSpec{}
+	fields := strings.Join(keys, ",")
+	query := fmt.Sprintf("SELECT %s FROM %s LIMIT 1", fields, meta.table)
+	sqlDB, _ := sql.Open("maxcompute", meta.odpsConfig.FormatDSN())
+	rows, err := sqlDB.Query(query)
+	if err != nil {
+		return output, err
+	}
+	defer sqlDB.Close()
+	columnTypes, _ := rows.ColumnTypes()
+	columns, _ := rows.Columns()
+	count := len(columns)
+	for rows.Next() {
+		values := make([]interface{}, count)
+		for i, ct := range columnTypes {
+			v, e := createByType(ct.ScanType())
+			if e != nil {
+				return output, e
+			}
+			values[i] = v
+		}
+		if err := rows.Scan(values...); err != nil {
+			return output, err
+		}
+		for _, ct := range columnTypes {
+			denseValue := values[0].(*string)
+			fields := strings.Split(*denseValue, ",")
+			shape := make([]int, 1)
+			shape[0] = len(fields)
+			output[ct.Name()] = &columnSpec{ct.Name(), false, false, shape, "float", ",", *meta.featureMap}
+		}
+	}
+	return output, nil
+}
+
+func (meta *Metadata) GetSparseColumnInfo() (map[string]*columnSpec, error) {
+	output := map[string]*columnSpec{}
+
+	filter := "feature_type != '' "
+	if meta.featureMap.Partition != "" {
+		filter += "and " + meta.featureMap.Partition
+	}
+	query := fmt.Sprintf("SELECT feature_type, max(cast(id as bigint)) as feature_num, group "+
+		"FROM %s WHERE %s GROUP BY group, feature_type", meta.featureMap.Table, filter)
+
+	sqlDB, _ := sql.Open("maxcompute", meta.odpsConfig.FormatDSN())
+	rows, err := sqlDB.Query(query)
+	if err != nil {
+		return output, err
+	}
+	defer sqlDB.Close()
```

Comment:
`defer sqlDB.Close()` should follow `sql.Open(...)`

Reply:
i think `defer sqlDB.Close()` should not be called if error occurs ?
Encouraged by:
https://github.com/go-sql-driver/mysql/wiki/Examples#a-word-on-sqlopen